### ISSUE TITLE
Support installing Slicer through Apple App Store

### ISSUE THREAD

**Issue by jcfr** [Slicer/Slicer#6065]
_This issue was created following discussion with @rkikinis. This issue will serve as documentation and placeholder for developers looking into addressing this_

## Is your feature request related to a problem? Please describe.

Apple users are expecting application to be available on the App Store

## Describe the solution you'd like

Searching for `3D Slicer` or `Slicer` on the app store should list the application and allow user to install it.

## Describe alternatives you've considered

Download from https://download.slicer.org/  

This is the approach currently used.

## Additional context

See https://developer.apple.com/app-store/review/guidelines/

**Comment by jcfr** [comment]
As of 2021.12.03, the section `2.5 Software Requirements`  of the [guidelines](https://developer.apple.com/app-store/review/guidelines/) indicates the following:

> 2.5.2 Apps should be self-contained in their bundles, and may not read or write data outside the designated container area, **nor may they download, install, or execute code which introduces or changes features or functionality of the app**, including other apps. Educational apps designed to teach, develop, or allow students to test executable code may, in limited circumstances, download code provided that such code is not used for other purposes. Such apps must make the source code provided by the app completely viewable and editable by the user.

For this reason, until we reach out to an Apple representative to further clarify or ask for an exemption, I am closing the issue.

**Comment by lassoan** [comment]
Does Chrome browser available on the Apple App Store? Can it install plugins?

**Comment by lassoan** [comment]
It seems that the limitation is only for **automatically** downloaded content. If the plugin install is a user action then it should be fine. There are apparently many applications with plugins infrastructure.

https://stackoverflow.com/questions/5319982/mac-app-store-and-plugins

> The [Mac Developer Agreement](https://developer.apple.com/programs/terms/mac/mac_program_agreement_20120216.pdf) at point 3.3.2 states
>- An Application may install or run additional interpreted or executable code (e.g., plug-ins and extensions) for use in conjunction with the Application as long as such code:
>- does not change the Application’s submitted binary or would not otherwise be considered an Update (as determined in Apple’s sole discretion); and
>- does not change the primary purpose of the Application by providing features or functionality that are inconsistent with the intended and advertised purpose of the Application as submitted to the Mac App Store.

**Comment by lassoan** [comment]
I don't think we can close this issue, because it should be addressed. I just move it to the backlog.

**Comment by lassoan** [comment]
By the way, Slicer on the appstore makes sense even if enabling the extension manager requires extra work. We could bundle the 10-15 most frequently used extensions (or even all extensions except the large or complicated ones).

**Comment by jcfr** [comment]
After attempting to install extension (e.g `SlicerHeart`) using the signed & notarized 5.0.2 macOS package, errors like the following are reported after installing the extension and restarting Slicer:

```
/path/to/Slicer-5.0.2.app/Contents/Extensions-30822/SlicerHeart/lib/Slicer-5.0/qt-loadable-modules/libqSlicerKretzFileReaderModule.dylib:
  code signature in (/path/to/Slicer-5.0.2.app/Contents/Extensions-30822/SlicerHeart/lib/Slicer-5.0/qt-loadable-modules/libqSlicerKretzFileReaderModule.dylib) not valid for use in process using Library Validation: mapped file has no cdhash, completely unsigned? Code has to be at least ad-hoc signed.)
```

Similarly, after forcing the re-install of numpy and restarting: 

```python
slicer.util.pip_install(["install", "-U", "numpy"])
slicer.util.restart()
```

messages like the following are reported:

```
Original error was: dlopen(/path/to/Slicer-5.0.2.app/Contents/lib/Python/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so, 2): no suitable image found.  Did find:
	/path/to/Slicer-5.0.2.app/Contents/lib/Python/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so: code signature in (/path/to/Slicer-5.0.2.app/Contents/lib/Python/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so) not valid for use in process using Library Validation: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?)
```

## Proposed solution

To address this, the current list of entitlement should be updated to include [com.apple.security.cs.disable-library-validation](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_disable-library-validation):

```diff
<plist version="1.0">
  <!--
  -->
  <dict>
    <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
    <true/>
  </dict>
+  <dict>
+    <key>com.apple.security.cs.disable-library-validation</key>
+    <true/>
+  </dict>
</plist>
```

See https://developer.apple.com/forums/thread/128380

## Impact on Apple App Store distribution

Assuming signing (without notarization) macOS extensions allows to successfully load them, the Slicer application through the app store would be able to install signed extension but wouldn't be able to install arbitrary python packages providing compiled dependencies.

**Comment by lassoan** [comment]
> Assuming signing (without notarization) macOS extensions allows to successfully load them, the Slicer application through the app store would be able to install signed extension but wouldn't be able to install arbitrary python packages providing compiled dependencies.

Thank you for exploring this. This is useful information and actually makes sense. It would be hard to ensure the integrity of binaries in Python packages that the user or an extension downloads. Probably similar restrictions apply to Microsoft Store as well. I guess we could still add signature to compiled Python binaries that we bundle in extension packages.

**Comment by jamesobutler** [comment]
https://doc.qt.io/qt-5/qtwebengine-platform-notes.html#mac-app-store-compatibility

https://doc.qt.io/qt-6/qtwebengine-platform-notes.html#mac-app-store-compatibility

> Applications using Qt WebEngine are not compatible with the Mac App Store, because:
> 
> - The Chromium part of the code uses several private API methods, which are prohibited by the App Store.
> - Applications submitted to the App Store must be code-signed with the App Sandbox feature enabled. The App Sandbox feature interferes with Chromium's own sandbox initialization, which results in Chromium not being properly initialized. This also ties in with the private API usage. Furthermore, standalone Chromium itself is not officially tested with the App Sandbox enabled, and even if work is done to bypass the App Store's restrictions, that would not guarantee correct behavior of the library.

**Comment by lassoan** [comment]
Thank you, this is useful information. 

The app store incompatibility may go away anytime. However, even if it stays it is not a blocking issue. 

Qt webengine is not essential. We could use the manage extension tab to browse and install all extensions with a few small changes. Alternatively, we could use the system web browser and use custom protocol handier (e.g., slicerextension://something.slicer.org/id=someid) to initiate installation of the selected extension.
